Implement a fingerprinting mechanism to track compaction states in a more efficient manner

capistrant opened the pull request:
<!-- Thanks for trying to help us make Apache Druid be the best it can be! Please fill out as much of the following information as is possible (where relevant, and remove it when irrelevant) to help make the intention and scope of this PR clear in order to ease review. -->

<!-- Please read the doc for contribution (https://github.com/apache/druid/blob/master/CONTRIBUTING.md) before making this PR. Also, once you open a PR, please _avoid using force pushes and rebasing_ since these make it difficult for reviewers to see what you've changed in response to their reviews. See [the 'If your pull request shows conflicts with master' section](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#if-your-pull-request-shows-conflicts-with-master) for more details. -->

<!-- Replace XXXX with the id of the issue fixed in this PR. Remove this section if there is no corresponding issue. Don't reference the issue in the title of this pull-request. -->

<!-- If you are a committer, follow the PR action item checklist for committers:
https://github.com/apache/druid/blob/master/dev/committer-instructions.md#pr-and-issue-action-item-checklist-for-committers. -->

**disclaimer: This PR begins the migration away from always using the term compaction. It may create some short term friction/frustration. But the goal is that longer term we refer to things that are not pure compaction in a more ambiguous way (indexing state versus compaction state, for example)**

### Description

Add new functionality to Compaction Supervisors. Instead of storing compaction state for segments individually, centralize the state storage in a new metadata table. Individual segments will store a computed fingerprint that references an indexing state (aka compaction state) in the new metadata table. Since many segments will eventually end up sharing common indexing states, this should greatly reduce duplication in metadata storage.

**note:** This applies only to Compaction Supervisors. Scheduled compaction on the coordinator will not use fingerprinting. This is intentional as a part of the roadmap to sunsetting coordinator duty compaction in the future.

<!-- Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's not necessary to repeat the description here, however, you may choose to keep one summary sentence. -->

<!-- Describe your patch: what did you change in code? How did you fix the problem? -->

<!-- If there are several relatively logically separate changes in this PR, create a mini-section for each of them. For example: -->

####  Indexing State Fingerprinting

Instead of storing serialized `CompactionState` objects as the `lastCompactionState` field in every compacted segment, generate a fingerprint for a `CompactionState` and attach that to compacted segments. Add new centralized storage for serialized `CompactionState` objects where individual states can be looked up by the aforementioned fingerprint. Since it is common for many segments in a data source to share a single `CompactionState`, this greatly reduces the metadata storage overhead for storing indexing states in the metadata store.

#####  Metadata Store Changes

###### `druid_segments`

Add new column `indexing_state_fingerprint` that stores the fingerprint representation of the segments current `CompactionState`. It can be `null` if no compaction has taken place.

###### `druid_indexingStates`

New metadata table that stores the full serialized `CompactionState` associated with a fingerprint. Segments can look up their detailed state here by using the `indexing_state_fingerprint` that they are associated with to pull the full state payload.

##### `IndexingStateStorage`

###### `SqlIndexingStateStorage`

The `IndexingStateStorage` interface + implementation(s) is responsible for managing the persistence and lifecycle of indexing states. It stores unique compaction configurations (identified by fingerprints) in the metadata database. The manager tracks which compaction states are actively referenced by segments, marking unreferenced states as unused and periodically cleaning up old unused states. This fingerprinting approach allows Druid to efficiently store and retrieve compaction metadata without duplicating identical compaction configurations across multiple segments.

###### `HeapMemoryIndexingStateStorage`

Meant to serve as a mechanism for testing and simulations where metadata persistence may not be available/needed

##### `IndexingStateCache`

`IndexingStateCache` is a new component of the `HeapMemorySegmentMetadataCache`. It is modeled strongly after the existing datasource schema cache. This is where the existing indexing states are cached for reference by compaction supervisors.

#### CompactSegments Coordinator Duty Roadmap

This PR does not add support for indexing state fingerprinting to the coordinator based scheduled compaction that is carried out by the CompactSegments coordinator duty. This is because the Druid roadmap is to move all scheduled compaction to compaction supervisors. Making the decision to forgo indexing state fingerprint support for the legacy duty based compaction is a conscious choice we are making to help drive usage of supervisors and limit changes to the legacy duty based compaction code. Another PR should be spun up to officially deprecate legacy scheduled compaction on the coordinator. 

#### Legacy `lastCompactionState` Roadmap

This PR implements no automatic transition to fingerprints for segments who are compacted and store `CompactionState` in their `lastCompactionState` field. Instead this PR aims to continue supporting `lastCompactionState` in Compaction decision making for segments compacted before fingerprinting. This means that legacy segments **will not** have to be re-compacted simply because they are not fingerprinted, as long as they have the proper `CompactionState` as specified by the compaction configuration for the data source in question.

This PR also writes both the new fingerprint as well as the legacy `lastCompactionState`  to a `DataSegment` by default. This allows normal rolling upgrade order as well as Druid version rollback without un-needed re-compaction. An operator can disable writing `lastCompactionState` by updating the cluster compaction config, after the Druid upgrade completes. Eventually, Druid code base will cease writing `lastCompactionState` at all and instead force using fingerprinting going forward. I think this should be done in the Druid version following the first version that this new feature is seen in. Even at this point, `lastCompactionState` will need to continue to be supported for already written segments, unless we want to devise an automated migration plan that can run in the background of a cluster to get all compacted segments migrated to fingerprinting.

#### Follow ups needed

* Refactor existing `CompactionState` references in the app code to IndexingState
  * other than `lastCompactionState` in data segments and metastore which will need to be carried onwards. 
  * This PR focused only on the naming migration in new code.
* Update automatic compaction documentation
  * Deprecate automatic compaction on the coordinator
  * Mark automatic compaction with compaction supervisors as GA
  * Create a migration guide for folks going from legacy coordinator compaction to compaction supervisors
    * Note that we are still rollback compatible in terms of going from supervisors --> coordinator automatic compaction, without re-triggering compaction unnecessarily.

<!--
In each section, please describe design decisions made, including:
 - Choice of algorithms
 - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such as when there are insufficient resources?
 - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
 - Method organization and design (how the logic is split between methods, parameters and return types)
 - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
-->


<!-- It's good to describe an alternative design (or mention an alternative name) for every design (or naming) decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen) to highlight the advantages of the chosen designs and names. -->

<!-- If there was a discussion of the design of the feature implemented in this PR elsewhere (e. g. a "Proposal" issue, any other issue, or a thread in the development mailing list), link to that discussion from this PR description and explain what have changed in your final design compared to your original proposal or the consensus version in the end of the discussion. If something hasn't changed since the original discussion, you can omit a detailed discussion of those aspects of the design here, perhaps apart from brief mentioning for the sake of readability of this PR description. -->

<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

#### Release note
<!-- Give your best effort to summarize your changes in a couple of sentences aimed toward Druid users. 

If your change doesn't have end user impact, you can skip this section.

For tips about how to write a good release note, see [Release notes](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#release-notes).

-->

Improvements to automatic compaction supervisors. Instead of individual compacted segments having to store their full compaction state in the metadata store, states will now be stored in a central location in the metadata store ( a new `indexingStates` table). Individual segments will only need to store a unique reference (indexing_state_fingerprint) to their full compaction state. Since it is typical that many segments in a single datasource share the same underlying compaction state, this will end up greatly reducing metadata storage requirements for automatic compaction.

For backward compatibility purposes, detailed compaction state will continue to be persisted in each segment until a future date. An operator can opt out of this by setting `storeCompactionStatePerSegment` to `false` in the cluster compaction config. This should only be done with the understanding that rolling back to an earlier version of Druid would trigger re-compaction on any segments who had been compacted since updating that config.

It must also be noted that using Automatic Compaction Supervisors now requires incremental segment metadata caching to be enabled on the Overlord and Coordinator via runtime properties.

##### Upgrade Note

Automatic Compaction Supervisors now requires incremental segment metadata caching to be enabled on the Overlord and Coordinator via runtime properties. `druid.manager.segments.useIncrementalCache` must not be `never`. See configuration [docs](https://druid.apache.org/docs/latest/configuration/#metadata-retrieval) for details.

Metadata store changes are required for this upgrade. If you already have `druid.metadata.storage.connector.createTables` set to `true` no action is needed. If you have this feature disabled, you will need to alter the `segments` table and create the `compactionStates` table. Postgres DDL is provided below as a guide. You will have to adapt the syntax to your metadata store backend as well as use proper table naming depending on your configured table prefix and database.

```SQL
-- create the indexing states lookup table and associated indices
CREATE TABLE druid_indexingStates (
    created_date VARCHAR(255) NOT NULL,
    datasource VARCHAR(255) NOT NULL,
    fingerprint VARCHAR(255) NOT NULL,
    payload BYTEA NOT NULL,
    used BOOLEAN NOT NULL,
    pending BOOLEAN NOT NULL,
    used_status_last_updated VARCHAR(255) NOT NULL,
    PRIMARY KEY (fingerprint),
  );

  CREATE INDEX idx_druid_compactionStates_used ON druid_compactionStates(used, used_status_last_updated);
```

```SQL
-- modify druid_segments table to have a column for storing compaction state fingerprints
ALTER TABLE druid_segments ADD COLUMN indexing_state_fingerprint VARCHAR(255);
```


<hr>

##### Key changed/added classes in this PR
 * `CompactionStatus`
 * `CompactionConfigBasedJobTemplate`
 * `CompactionState`
 * `SQLMetadataConnector`
 * `IndexingStateStorage`
 * `SqlIndexingStateStorage`
 * `IndexingStateCache`
 * `CompactSegments`
 * `KillUnreferencedIndexingState`

<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

capistrant commented on indexing-service/src/main/java/org/apache/druid/indexing/compact/CompactionConfigBasedJobTemplate.java:
```diff
@@ -70,12 +73,32 @@ public List<CompactionJob> createCompactionJobs(
+        config.getDataSource()
+    );
+
+    if (segmentIterator.hasNext()) {
```
comment here may be useful. Clear up that we pre-write the compaction state we will be creating compaction tasks for so the underlying task does not need to worry about it, all it has to do is persist the fingerprint when it generates segments.

capistrant commented on server/src/main/java/org/apache/druid/server/compaction/CompactionStatus.java:
```diff
@@ -376,15 +416,92 @@ private CompactionStatus evaluate()
+     * {@link #unknownStateToSegments} where their compaction states will be analyzed.
+     * </p>
+     */
+    private CompactionStatus allFingerprintedCandidatesHaveExpectedFingerprint()
```
i think this is missing adding properly fingerprinted segments to compacted segments which is used for stats?

capistrant commented on server/src/main/java/org/apache/druid/segment/metadata/CompactionStateManager.java:
```diff
@@ -0,0 +1,538 @@
+import java.util.concurrent.locks.Lock;
+
+/**
+ * Handles compaction state persistence on the Coordinator.
```
better java doc. this isn't accurate either. if using supervisors, the overlord manages state

capistrant commented on server/src/main/java/org/apache/druid/segment/metadata/CompactionStateManager.java:
is an associated test class missing?

capistrant commented on server/src/main/java/org/apache/druid/segment/metadata/CompactionStateManager.java:
what about if the operator has create tables disabled and does not properly create the table before upgrading?

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/HeapMemoryCompactionStateManager.java:
```diff
@@ -0,0 +1,81 @@
+ * In-memory implementation of {@link CompactionStateManager} that stores
+ * compaction state fingerprints in heap memory without requiring a database.
+ * <p>
+ * Useful for simulations and unit tests where database persistence is not needed.
```
If this is used only in tests, we should probably put it in the test source root `src/test/java`.

kfaraz replied:
Oh, I see. Are you referring to `CoordinatorSimulationBuilder` or some other class?

kfaraz commented on processing/src/main/java/org/apache/druid/timeline/CompactionState.java:
```diff
@@ -50,6 +59,26 @@
 {
+
+  /**
+   * Lazy initialization holder for deterministic ObjectMapper.
```
I wonder if we shouldn't just inject this mapper annotated with `@Sorted` or `@Deterministic` as a lazy singleton. It may be injected into `CompactionStateManager` and fingerprints will always be created by that class rather than using a static utility method.

kfaraz commented on processing/src/main/java/org/apache/druid/metadata/MetadataStorageConnector.java:
```diff
@@ -95,4 +95,9 @@ default void exportTable(
   void createSegmentSchemasTable();
+
+  /**
+   *
```
Doc missing?

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/ClusterCompactionConfig.java:
```diff
@@ -45,21 +45,34 @@ public class ClusterCompactionConfig
+   * Instead, it will only be stored in the metadata store with a fingerprint id that segments will reference. Some
+   * operators may want to disable this behavior early to begin saving space in segment metadatastore table entries.
+   */
+  private final boolean legacyPersistLastCompactionStateInSegments;
```
Does this need to be a dynamic config? Feels like it should be a runtime property.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/compact/CompactionConfigBasedJobTemplate.java:
```diff
@@ -70,12 +73,34 @@ public List<CompactionJob> createCompactionJobs(
+    if (segmentIterator.hasNext()) {
+      // If we are going to create compaction jobs for this compaction state, we need to persist the fingerprint -> state
+      // mapping so compacted segments from these jobs can reference a valid compaction state.
+      params.getCompactionStateManager().persistCompactionState(
```
The templates should only perform lightweight (i.e. non-IO) read-only operations as `createCompactionJobs` may be called frequently.
We should not do any persistence here.
Instead, the `params` can hold some mapping where we can add this compaction state and perform persistence on-demand (perhaps in the `CompactionJobQueue`).

capistrant replied:
thank you for the guidance. Will work on how to get this out of hot path

kfaraz commented on website/.spelling:
```diff
@@ -483,6 +483,7 @@ pre-computation
 pre-compute
 pre-computed
 pre-computing
+pre-dates
```
`predates` need not be hyphenated.

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/CompactionStateManager.java:
```diff
@@ -0,0 +1,545 @@
+ * </p>
+ */
+@ManageLifecycle
+public class CompactionStateManager
```
I don't feel that pre-warming the cache is really necessary. The fingerprint needs to be retrieved only when running the `CompactionJobQueue` on Overlord or `CompactSegments` on Coordinator.

1. Let's always keep all the compaction states in memory. We are already keeping all the used segments in memory. The distinct `CompactionState` objects will be fairly small in number and size.
2. The states can be cached in `HeapMemorySegmentMetadataCache` which already serves as a cache for used segments, pending segments and schemas.
3. If possible, let's support the fingerprint flow only with compaction supervisors and not the Coordinator-based `CompactSegments` duty. That would simplify the new flow and be another motivation for users to migrate to using compaction supervisors.

capistrant replied:
> If possible, let's support the fingerprint flow only with compaction supervisors and not the Coordinator-based CompactSegments duty. That would simplify the new flow and be another motivation for users to migrate to using compaction supervisors.

would we want to deprecate `CompactSegments` compaction on the coordinator in this case? so we aren't forever supporting compaction without fingerprints + compaction with fingerprints?

kfaraz replied:
Yes, the plan was to deprecate `CompactSegments` once compaction supervisors took off. I don't fully recall if compaction supervisors is already marked GA or not. They would also have to be made the default, if we want to start deprecation of `CompactSegments`.

But I feel all of this should be out of scope for the current PR.

If supporting the fingerprint logic in `CompactSegments` is not additional work and does not complicate the flow, we can leave it as is.

My only concern is that there should be just one service that is responsible for persisting new fingerprints. I would prefer that to be the Overlord, so that it always has a consistent cache state. So we either just don't support fingerprints on the Coordinator or we handle persistence by calling an Overlord API.

(I am yet to go through the whole PR to identify all the call sites that may persist a compaction state. I have only found the one in `CompactionConfigBasedJobTemplate` so far.)

kfaraz reviewed (commented):
Thanks for the feature, @capistrant !

I have started going through the PR, leaving a partial review here.
I am yet to go through several changes, such as the ones made in `CompactionStatus`, `DatasourceCompactibleSegmentIterator`, etc.

capistrant commented on processing/src/main/java/org/apache/druid/timeline/DataSegment.java:
```diff
@@ -245,6 +256,9 @@ public DataSegment(
     this.size = size;
+    this.compactionStateFingerprint = compactionStateFingerprint != null
+                                      ? STRING_INTERNER.intern(compactionStateFingerprint)
+                                      : null;
```
also calling intern on a null is an NPE from what I can tell testing this locally

capistrant commented on server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java:
```diff
@@ -1701,6 +1705,123 @@ private SegmentSchemaRecord mapToSchemaRecord(ResultSet resultSet)
+                .collect(Collectors.toList());
+  }
+
+  private CompactionStateRecord mapToCompactionStateRecord(ResultSet resultSet)
```
javadoc

capistrant commented on server/src/main/java/org/apache/druid/metadata/segment/cache/HeapMemorySegmentMetadataCache.java:
```diff
@@ -265,6 +273,9 @@ public void stopBeingLeader()
   {
     synchronized (cacheStateLock) {
       if (isEnabled()) {
+        if (useCompactionStateCache) {
```
other caches don't clear, why should we?

capistrant commented on server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java:
```diff
@@ -461,4 +487,59 @@ private static ClientCompactionTaskQuery compactSegments(
+  /**
+   * Given a {@link DataSourceCompactionConfig}, create a {@link CompactionState}
+   */
+  public static CompactionState createCompactionStateFromConfig(DataSourceCompactionConfig config)
```
Is this still the right place for this now that we only support for supervisors?

capistrant commented on server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnreferencedCompactionState.java:
```diff
@@ -0,0 +1,61 @@
+
+import java.util.List;
+
+public class KillUnreferencedCompactionState extends MetadataCleanupDuty
```
javadoc

capistrant commented on server/src/test/java/org/apache/druid/server/coordinator/duty/KillUnreferencedCompactionStateTest.java:
```diff
@@ -0,0 +1,272 @@
+    return usedStatus.isEmpty() ? null : usedStatus.get(0);
+  }
+
+  private static ObjectMapper createDeterministicMapper()
```
does this need to react to the new way we are creating the mapper?

capistrant commented on services/src/main/java/org/apache/druid/guice/MetadataManagerModule.java:
```diff
@@ -123,6 +132,15 @@ public void configure(Binder binder)
             .in(LazySingleton.class);
     }
 
+    // Overlord-only compaction state dependencies
```
nit: maybe move this into one conditional for overlord only with stuff below?

kfaraz commented:
Thanks for the update, @capistrant ! 
I will try to finish a review of the latest changes today.

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/CompactionStateCache.java:
```diff
@@ -0,0 +1,147 @@
+   */
+  public void clear()
+  {
+    publishedCompactionStates.set(PublishedCompactionStates.EMPTY);
```
Should we also reset the hit and miss counts here?

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/CompactionStateCache.java:
```diff
@@ -0,0 +1,147 @@
+  public void clear()
+  {
+    publishedCompactionStates.set(PublishedCompactionStates.EMPTY);
+    log.info("Cleared compaction state cache");
```
I think we can omit this log line assuming that `clear()` is called only on service shutdown.

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/PersistedCompactionStateManager.java:
```diff
@@ -0,0 +1,401 @@
+ * </p>
+ */
+@ManageLifecycle
+public class PersistedCompactionStateManager implements CompactionStateManager
```
Nit: We may rename this to `SqlCompactionStateManager` to align with other similar classes that handle persistence of some metadata in the DB.

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/PersistedCompactionStateManager.java:
```diff
@@ -0,0 +1,401 @@
+      return;
+    }
+
+    final Lock lock = datasourceLocks.get(dataSource);
```
IIUC, we are using the lock so that the same fingerprint is not updated/inserted in the metadata store by another thread while we are working here.

I think we should do the following instead:
- We can remove the locking, since the persistence will happen from only a single thread anyway (via `CompactionJobQueue.runReadyJobs()`).
- We can just log an error, if there are conflicts since this method should be idempotent anyway.

capistrant replied:
yes the intention was to provide protection against unique constraint exceptions. I agree that it is over-engineered for the system as it currently is. When I was weighing alternative of trying to catch and ignore the unique constraint exceptions I didn't like how it looked hacky. I will revert back to that though as I get that this added complexity is probably worse.

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/CompactionStateManager.java:
```diff
@@ -0,0 +1,105 @@
+   * @param fingerprintToStateMap Map of fingerprints to their compaction states
+   * @param updateTime The timestamp for this update
+   */
+  void persistCompactionState(
```
The only usage of this method seems to pass in a single fingerprint right now. Should we just keep that for the time being. We can add the batch upsert when needed later.

capistrant replied:
ya, this is over-engineered for batching. I will simplify.

kfaraz commented on server/src/test/java/org/apache/druid/server/coordinator/duty/CompactSegmentsTest.java:
```diff
@@ -273,6 +276,7 @@ public void testSerde() throws Exception
             .addValue(DruidCoordinatorConfig.class, COORDINATOR_CONFIG)
             .addValue(OverlordClient.class, overlordClient)
             .addValue(CompactionStatusTracker.class, statusTracker)
+            .addValue(CompactionStateManager.class, compactionStateManager)
```
I think this can be removed now since `CompactSegments` doesn't use this anymore.

kfaraz commented on server/src/main/java/org/apache/druid/metadata/segment/cache/CompactionStateRecord.java:
```diff
@@ -0,0 +1,47 @@
+import org.apache.druid.timeline.CompactionState;
+
+/**
+ * Represents a single record in the compaction_states table.
```
```suggestion
 * Represents a single record in the druid_compactionStates table.
```

kfaraz commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -355,6 +355,7 @@ public void createSegmentTable(final String tableName)
     columns.add("used BOOLEAN NOT NULL");
     columns.add("payload %2$s NOT NULL");
     columns.add("used_status_last_updated VARCHAR(255) NOT NULL");
+    columns.add("compaction_state_fingerprint VARCHAR(255)");
```
Could you please also add the `upgraded_from_segment_id` column here? I think we had missed adding it, and it is present only in the `alterSegmentTable` method.

kfaraz commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -1097,6 +1100,55 @@ public void createSegmentSchemasTable()
+                + "  used BOOLEAN NOT NULL,\n"
+                + "  used_status_last_updated VARCHAR(255) NOT NULL,\n"
+                + "  PRIMARY KEY (id),\n"
+                + "  UNIQUE (fingerprint)\n"
```
If the fingerprint is unique anyway, do we still need an `id` column?
I forget why we had added an `id` column in the schemas table too.

capistrant replied:
uh I guess it probably doesn't matter? Maybe just a force of habit for people creating rdbms tables to throw in a serial id PK without thinking.

kfaraz commented:
@techdocsmith , I have tried to answer your queries below.

> druid_segments
> Is druid necessary here? What other segments migth we distinguish the Druid segments from?

This is an existing convention in the Druid codebase. The default names of the metadata tables use the prefix `druid_` and use a camel-cased name. Howevers, users can always configure their cluster to use some other name.

> compaction_state_fingerprint
If this is a column within the druid_compactionStates, could it just be called fingerprint ?

This is the foreign key column in the `druid_segments` table which refers to the `fingerprint` column in the `druid_compactionStates` table.

kfaraz commented on server/src/main/java/org/apache/druid/server/compaction/CompactionStatus.java:
```diff
@@ -351,40 +388,143 @@ private CompactionStatus evaluate()
-          .collect(Collectors.toList());
+      // Includes segments with correct fingerprints and segments that passed all state checks
+      final List<DataSegment> allCompactedSegments = new ArrayList<>(this.compactedSegments);
+      allCompactedSegments.addAll(
```
Nit: It might be cognitively simpler to just add the unknown state segments directly to `this.compactedSegments` instead of using a local variable `allCompactedSegments`.

kfaraz commented on server/src/main/java/org/apache/druid/server/compaction/CompactionStatus.java:
```diff
@@ -351,40 +388,143 @@ private CompactionStatus evaluate()
+
+      boolean fingerprintedSegmentNeedingCompactionFound = false;
+
+      if (compactionStateCache != null) {
```
Please simplify this condition to have a return early syntax.

e.g.

```java
if (cache == null) {
   // move all mismatching segments to unknown state
   return CompactionStatus.pending("Segments have a mismatching fingerprint");
}

// The rest of the logic goes here. Do not put in an else block to avoid unnecessary indentation
```

Just to confirm, the `compactionStateCache` can be null only when running the legacy `CompactSegments` duty on the Coordinator, right? It is never expected to be null when running supervisors on the Overlord?

capistrant replied:
> Just to confirm, the compactionStateCache can be null only when running the legacy CompactSegments duty on the Coordinator, right? It is never expected to be null when running supervisors on the Overlord?

correct.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/ClusterCompactionConfig.java:
```diff
@@ -45,21 +45,34 @@ public class ClusterCompactionConfig
       @JsonProperty("useSupervisors") @Nullable Boolean useSupervisors,
-      @JsonProperty("engine") @Nullable CompactionEngine engine
+      @JsonProperty("engine") @Nullable CompactionEngine engine,
+      @JsonProperty("legacyPersistLastCompactionStateInSegments") Boolean legacyPersistLastCompactionStateInSegments
```
Can we call this field `storeCompactionStatePerSegment` instead?
Then it would align with the task context parameter, `storeCompactionState`.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/compact/CompactionJob.java:
```diff
@@ -31,27 +32,37 @@ public class CompactionJob extends BatchIndexingJob
   private final CompactionCandidate candidate;
   private final int maxRequiredTaskSlots;
+  private final String compactionStateFingerprint;
+  private final CompactionState compactionState;
```
Should we rename these to `targetCompactionState` and `targetCompactionStateFingerprint`?

capistrant commented on server/src/main/java/org/apache/druid/segment/metadata/CompactionStateManager.java:
```diff
@@ -0,0 +1,105 @@
+
+  /**
+   * Marks compaction states as unused if they are not referenced by any used segments.
+   * This is used for cleanup operations. Implementations may choose to no-op this.
```
Fair point, I'll remove default impl and re-word doc.

capistrant commented on server/src/main/java/org/apache/druid/metadata/segment/cache/HeapMemorySegmentMetadataCache.java:
```diff
@@ -785,13 +800,13 @@ private void retrieveAllUsedSegments(
     if (useSchemaCache) {
       sql = StringUtils.format(
-          "SELECT id, payload, created_date, used_status_last_updated, schema_fingerprint, num_rows"
+          "SELECT id, payload, created_date, used_status_last_updated, compaction_state_fingerprint, schema_fingerprint, num_rows"
```
Hm. I intentionally did this to make all of the indexes static. Putting compaction_state_fingerprint at the end would result in its own index being dynamic which felt worse than a re-shuffle. Thoughts?

capistrant commented on server/src/main/java/org/apache/druid/metadata/segment/cache/HeapMemorySegmentMetadataCache.java:
```diff
@@ -1106,6 +1122,89 @@ private void emitMetric(String datasource, String metric, long value)
+        sql -> sql.retrieveCompactionStatesForFingerprints(addedFingerprints)
+    );
+    if (addedCompactionStateRecords.size() < addedFingerprints.size()) {
+      emitMetric(Metric.SKIPPED_COMPACTION_STATES, addedFingerprints.size() - addedCompactionStateRecords.size());
```
sure, I kinda fell trap to just replicating what the schema cache metric was. You want to delete this one or just add the added and deletes on top of this?

capistrant commented on server/src/test/java/org/apache/druid/segment/metadata/HeapMemoryCompactionStateManager.java:
```diff
@@ -0,0 +1,151 @@
+   * Creates an ObjectMapper configured for deterministic serialization.
+   * Used for generating consistent fingerprints.
+   */
+  private static ObjectMapper createDeterministicMapper()
```
not clear to me where you are indicating. I see TestUtils in indexing-service module which isn't visible from here. Plus there is TestUtils in processing module that is visible, but that is for cgroups metrics stuff and feels like not what you are indicating.

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/SqlCompactionStateStorage.java:
```diff
@@ -398,4 +313,40 @@ private static void bindValuesToInClause(
+        // SQL standard unique constraint violation codes
+        // 23505 = unique_violation (PostgreSQL, Derby)
+        // 23000 = integrity_constraint_violation (MySQL and others)
+        if ("23505".equals(sqlState) || "23000".equals(sqlState)) {
```
It would be cleaner to add the method `isUniqueConstraintViolation` method to `SQLMetadataConnector` so that each concrete DB may provide its own implementation, rather than handle all possible cases here.

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/SqlCompactionStateStorage.java:
```diff
@@ -94,140 +76,93 @@ public PersistedCompactionStateManager(
+        return;
+      }
+      // For other exceptions, let them propagate
+      throw e;
```
Nit: Probably cleaner to use if-else.
```java
if (unique constraint violated) {
   // log the details
} else {
   throw e;
}

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/DefaultCompactionFingerprintMapper.java:
```diff
@@ -0,0 +1,64 @@
+  @Override
+  public String generateFingerprint(String dataSource, CompactionState compactionState)
+  {
+    return compactionStateStorage.generateCompactionStateFingerprint(compactionState, dataSource);
```
I wonder if we need to delegate to `CompactionStateStorage` anymore.
Is it okay to just invoke the static utility in `CompactionFingerprints` here itself?
It would allow us to remove the `generateFingerprint` method from `CompactionStateStorage`, as it doesn't really fit there.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/compact/CompactionJobQueue.java:
```diff
@@ -125,8 +125,8 @@ public CompactionJobQueue(
-        compactionStateManager,
-        compactionStateCache
+        fingerprintMapper,
+        compactionStateStorage
```
Instead of passing the storage to the job params, we should probably just keep a reference to it in the `CompactionJobQueue` class itself since we only do the persistence here.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/compact/CompactionConfigBasedJobTemplate.java:
```diff
@@ -72,9 +71,9 @@ public List<CompactionJob> createCompactionJobs(
+    CompactionState compactionState = config.toCompactionState();
 
-    String compactionStateFingerprint = params.getCompactionStateManager().generateCompactionStateFingerprint(
+    String compactionStateFingerprint = params.getCompactionStateStorageImpl().generateCompactionStateFingerprint(
```
Use fingerprint mapper here instead.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/DataSourceCompactionConfig.java:
```diff
@@ -89,4 +98,59 @@ public interface DataSourceCompactionConfig
+  /**
+   * Converts this compaction config to a {@link CompactionState}.
+   */
+  default CompactionState toCompactionState()
```
Thanks for adding this!

kfaraz commented on docs/api-reference/automatic-compaction-api.md:
```diff
@@ -889,6 +889,7 @@ This includes the following fields:
 |`compactionPolicy`|Policy to choose intervals for compaction. Currently, the only supported policy is [Newest segment first](#compaction-policy-newestsegmentfirst).|Newest segment first|
 |`useSupervisors`|Whether compaction should be run on Overlord using supervisors instead of Coordinator duties.|false|
 |`engine`|Engine used for running compaction tasks, unless overridden in the datasource-level compaction config. Possible values are `native` and `msq`. `msq` engine can be used for compaction only if `useSupervisors` is `true`.|`native`|
+|`legacyPersistLastCompactionStateInSegments`|Whether to persist the full compaction state in segment metadata. When `true` (default), compaction state is stored in both the segment metadata and the compaction states table. This is historically how Druid has worked. When `false`, only a fingerprint reference is stored in the segment metadata, reducing storage overhead in the segments table. The a ...
```
Please update the config name.

kfaraz commented on server/src/main/java/org/apache/druid/server/compaction/CompactionStatus.java:
```diff
@@ -351,40 +382,138 @@ private CompactionStatus evaluate()
+      }
+
+      if (fingerprintedSegmentWithoutCachedStateFound) {
+        return CompactionStatus.pending("One or more fingerprinted segments do not have a cached compaction state");
```
Good catch! I just assumed that if a segment has a fingerprint, the corresponding state would always be available in the cache. But I guess there can be error scenarios where this happens. Should we also include a WARN log here since this shouldn't happen under normal conditions?

kfaraz commented on server/src/test/java/org/apache/druid/segment/metadata/HeapMemoryCompactionStateManager.java:
```diff
@@ -0,0 +1,151 @@
+      final String dataSource
+  )
+  {
+    final Hasher hasher = Hashing.sha256().newHasher();
```
Yes, it makes sense to use the static utility or use the `DefaultFingerprintMapper` directly in tests.

kfaraz commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -1097,6 +1100,55 @@ public void createSegmentSchemasTable()
+                "CREATE TABLE %1$s (\n"
+                + "  id %2$s NOT NULL,\n"
+                + "  created_date VARCHAR(255) NOT NULL,\n"
+                + "  datasource VARCHAR(255) NOT NULL,\n"
```
Ah, I see. thanks!

kfaraz reviewed (commented):
Thanks for addressing the feedback proactively, @capistrant !
The new code structure with the fingerprint mapper looks much better, and has made a couple of things apparent.

The major open items that need to be addressed are:
- Update the cache when upserting a fingerprint in the metadata store.
- Persist the target compaction state only when committing segments.
- Move cleanup duty to Overlord
- Rename new table and columns

capistrant commented on server/src/main/java/org/apache/druid/segment/metadata/SqlCompactionStateStorage.java:
```diff
@@ -94,140 +76,93 @@ public PersistedCompactionStateManager(
+              dataSource
           );
-          String setFingerprintsUsedSql = StringUtils.format(
+          String updateSql = StringUtils.format(
```
ya this was a big oversight and woulda been tons of useless update queries

capistrant commented on server/src/main/java/org/apache/druid/segment/metadata/SqlCompactionStateStorage.java:
```diff
@@ -94,140 +76,93 @@ public PersistedCompactionStateManager(
+                "Published compaction state for fingerprint[%s] to DB for datasource[%s].",
+                fingerprint,
                 dataSource
             );
```
ya great call out

capistrant commented on server/src/main/java/org/apache/druid/segment/metadata/SqlCompactionStateStorage.java:
```diff
@@ -94,140 +76,93 @@ public PersistedCompactionStateManager(
   {
-    if (fingerprintToStateMap.isEmpty()) {
+    if (compactionState == null || fingerprint == null || fingerprint.isEmpty()) {
       return;
```
ya makes sense to have more strict sanitization here around a db write. I feel like what I am pushing up is a bit verbose though

capistrant commented on indexing-service/src/main/java/org/apache/druid/indexing/compact/CompactionJobQueue.java:
```diff
@@ -357,14 +357,15 @@ private String startTaskIfReady(CompactionJob job)
-    if (job.getCompactionState() != null && job.getCompactionStateFingerprint() != null) {
-      jobParams.getCompactionStateManager().persistCompactionState(
+    if (job.getTargetCompactionState() != null && job.getTargetCompactionStateFingerprint() != null) {
+      jobParams.getCompactionStateStorageImpl().upsertCompactionState(
```
One solution I have in mind is a `pending` flag added to druid_compactionStates

when the job queue inserts states, it inserts them as pending=true. The kill duty then has a separate, more permissive, config for killing pending states. say p7d default. There should be very few states that get orphaned in a way that they ever need to be cleaned up this way. The config could suggest an operator never change this to a value less than the p999 runtime of their compaction tasks (and likely to just stick with default). Then the completion flow of a compaction task can also fire a query to update pending=false for its fingerprint in druid_compactionStates (most of the time this will be a 0 row update)

capistrant commented:
> > Thanks for addressing the feedback proactively, @capistrant ! The new code structure with the fingerprint mapper looks much better, and has made a couple of things apparent.
> > The major open items that need to be addressed are:
> > 
> > * Update the cache when upserting a fingerprint in the metadata store.
> > * Persist the target compaction state only when committing segments.
> > * Move cleanup duty to Overlord
> > * Rename new table and columns
> 
> Addressed 1 and 3
> 
> 2 - I have commented about the **why** behind this choice, but need to consider more the problems/risks you call out cuz they seem legit
> 
> 4 - I still plan to do last for purposes of the diff. Now that most all of the major things are sorted, I will stage commits for it locally

For 2, I ended up going with the idea of a state being `pending` when it it inserted in to the database by the job queue. The overlord cleanup duty uses a secondary cleanup mechanism for pending compaction states, allowing operators to be more aggressive about cleaning up the unused states while not risking a compaction state being nuked before a long running compaction task completes. When inserting new segments, the coordinator will flip any associated compaction states from pending to "active" (pending=false) if needed.

capistrant commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/duty/KillUnreferencedIndexingState.java:
```diff
@@ -0,0 +1,68 @@
+import javax.inject.Inject;
+import java.util.List;
+
+public class KillUnreferencedIndexingState extends OverlordMetadataCleanupDuty
```
javadoc

capistrant commented on indexing-service/src/test/java/org/apache/druid/indexing/compact/OverlordCompactionSchedulerTest.java:
```diff
@@ -231,7 +233,10 @@ private void initScheduler()
+        OBJECT_MAPPER,
+        new HeapMemoryIndexingStateStorage(),
+        new IndexingStateCache(),
+        OBJECT_MAPPER // TODO fix
```
fixup

capistrant commented on server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:
```diff
@@ -2685,6 +2713,46 @@ public Map<String, Set<String>> retrieveUpgradedToSegmentIds(
   }
 
+  /**
+   * Marks compaction state fingerprints as active (non-pending) for successfully published segments.
```
multiple parts of this comment need to update wording. method name needs refactor as well

capistrant commented on server/src/main/java/org/apache/druid/metadata/segment/cache/HeapMemorySegmentMetadataCache.java:
```diff
@@ -1106,6 +1122,89 @@ private void emitMetric(String datasource, String metric, long value)
   }
 
+  /**
+   * Retrieves required used compaction states from the metadata store and resets
```
commend and method name (and maybe code need refactor for naming)

capistrant commented on server/src/main/java/org/apache/druid/metadata/segment/cache/HeapMemorySegmentMetadataCache.java:
```diff
@@ -1106,6 +1122,89 @@ private void emitMetric(String datasource, String metric, long value)
+  }
+
+  /**
+   * Retrieves all used compaction states from the metadata store and builds a
```
indexing states

capistrant commented on server/src/test/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinatorTest.java:
```diff
@@ -4315,6 +4332,97 @@ public void testWriteOperation_alwaysUsesCache_inModeIfSynced()
   }
 
+  @Test
+  public void testCommitSegmentsAndMetadata_marksPendingCompactionStateAsActive()
```
name

clintropolis commented on processing/src/main/java/org/apache/druid/guice/DruidSecondaryModule.java:
```diff
@@ -164,6 +168,22 @@ public WireTransferableContext getWireTransferableContext(
+      Map<ByteBuffer, WireTransferable.Deserializer> wtDeserializers
+  )
+  {
+    final ObjectMapper sortedMapper = new DefaultObjectMapper();
```
is this cool? will this be missing all of the jackson modules that get registered with the normal jsonMapper or does it not matter?

kfaraz replied:
> is this cool? as in, like does it matter that this will this be missing all of the jackson modules that get registered with the normal jsonMapper?

Oh, wouldn't invoking `setupJackson` take care of that?
IIUC, the only thing we would miss is the `DefaultObjectMapper` being initialized with service name (as done in `JcaksonModule` for the other mappers). Please correct me if I am missing something.

> also, it seems like we inject it places so that we can make a DefaultIndexingStateFingerprintMapper, should this just be an internal implementation detail of DefaultIndexingStateFingerprintMapper? 

Yeah, I suppose this would be okay too. Although, we would still need to pass in the default `@Json ObjectMapper` and then make a copy inside the fingerprint mapper. But I agree that it would be less error prone.

> I would imagine in the future we would want to just get the fingerprint mapper from like the supervisor? (if it is configurable per datasource) or some fingerpint factory or something (if system wide) instead of this special object mapper used for the default impl in the future once this is made more pluggable unless i'm missing something

Hmm, I am not sure. The logic to generate a fingerprint for a given indexing state and to store and retrieve the state/fingerprint would continue to remain core Druid logic. Supervisors provided by extensions may just have their custom (serializable) implementations of the `CompactionState` class.

clintropolis replied:
>Oh, wouldn't invoking setupJackson take care of that?

I think that would setup the injectable values for `@JacksonInject`, but do we not need to register the jackson modules from all the druid modules with it like here? https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/guice/DruidInjectorBuilder.java#L223

It seems weird to add this mapper there though. Maybe this doesn't matter all that much because we never deserialize with this mapper, just make some bytes out of the compaction state... so it doesn't really matter as long as it serializes to a stable set of bytes?

clintropolis commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -229,6 +229,17 @@ protected boolean isRootCausePacketTooBigException(Throwable t)
+  public boolean isUniqueConstraintViolation(Throwable t)
+  {
+    return false;
+  }
```
should there be a default implementation for this? based on the usage, it looks like not implementing it means something will explodes instead of eating a (possibly expected?) exception

kfaraz replied:
I don't think the default impl would be universally applicable anyway.
So, we either keep this method abstract and force extensions to implement it, or let the transaction fail (as is currently being done in the patch).

Also, looking at the way we currently insert segments in `IndexerSQLMetadataConnector`, we just check once if a segment ID already exists and then skip the insert. Otherwise, we proceed with the insert and let the transaction fail if already inserted by a competing transaction. It is up to the caller to retry or swallow the exception in such cases.

I think it would be fine to do the same here too to keep things simple for the time being.

@capistrant , @clintropolis , thoughts?

clintropolis commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -355,6 +366,8 @@ public void createSegmentTable(final String tableName)
+    columns.add("indexing_state_fingerprint VARCHAR(255)");
+    columns.add("upgraded_from_segment_id VARCHAR(255)");
 
     if (centralizedDatasourceSchemaConfig.isEnabled()) {
```
i know this isn't new or yours, but it feels weird that we conditionally define the schema based on some config....

clintropolis commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -1243,12 +1300,12 @@ private void validateSegmentsTable()
       throw new ISE(
           "Cannot start Druid as table[%s] has an incompatible schema."
-          + " Reason: One or all of these columns [used_status_last_updated, schema_fingerprint, num_rows] does not exist in table."
+          + " Reason: One or all of these columns [used_status_last_updated, schema_fingerprint, num_rows, indexing_state_fingerprint] does not exist in table."
```
nit: might be about time to break this down and have separate messages per problem

clintropolis commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/duty/OverlordMetadataCleanupDuty.java:
```diff
@@ -0,0 +1,126 @@
+   *
+   * @return Number of deleted pending metadata entries
+   */
+  protected abstract int cleanupEntriesCreatedBeforePendingDurationToRetain(DateTime minCreatedTime);
```
same comment as ^, `cleanupPendingEntriesCreatedBefore` or `cleanupPendingEntriesOlderThan`

clintropolis commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/duty/KillUnreferencedIndexingState.java:
```diff
@@ -0,0 +1,79 @@
+  @Override
+  protected int cleanupEntriesCreatedBeforeDurationToRetain(DateTime minCreatedTime)
+  {
+    // 1: Mark unreferenced states as unused
```
current comments seem sort of redundant with the code, but it does seem like it would be useful to instead summarize why we are doing this other stuff before deleting unused

capistrant replied:
I nuked the comments and added a javadoc for the impl to provide context to what is going on with prepping the data before looking for rows to delete.

kfaraz commented on server/src/main/java/org/apache/druid/segment/metadata/SqlIndexingStateStorage.java:
```diff
@@ -0,0 +1,408 @@
+
+  @Inject
+  public SqlIndexingStateStorage(
+      @Nonnull MetadataStorageTablesConfig dbTables,
```
Nit: Do we need the `@Nonnull` annotations? I don't think we use them in injected constructors anyway.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/duty/OverlordMetadataCleanupDuty.java:
```diff
@@ -0,0 +1,126 @@
+   *
+   * @return Number of deleted metadata entries
+   */
+  protected abstract int cleanupEntriesCreatedBeforeDurationToRetain(DateTime minCreatedTime);
```
+1

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/OverlordMetadataCleanupConfig.java:
```diff
@@ -0,0 +1,100 @@
+
+import java.util.Objects;
+
+public class OverlordMetadataCleanupConfig
```
+1, @capistrant , when we migrate other cleanup duties to the Overlord, we will just use the `MetadataCleanupConfig` itself.
For the current purposes, let's add a new class `IndexingStateCleanupConfig` which extends `MetadataCleanupConfig`.

Would that work?

capistrant commented on indexing-service/src/main/java/org/apache/druid/indexing/compact/OverlordCompactionScheduler.java:
```diff
@@ -164,8 +162,7 @@ public OverlordCompactionScheduler(
       ServiceEmitter emitter,
       ObjectMapper objectMapper,
```
should this be annotated with `@json`? @clintropolis @kfaraz

capistrant replied:
ok cool

capistrant commented on server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:
```diff
@@ -2729,25 +2729,23 @@ private void markIndexingStateFingerprintsAsActive(Set<DataSegment> segments)
-        // Log but don't fail the overall operation - the fingerprint will stay pending
-        // and be cleaned up by the pending grace period
-        log.warn(e, "Failed to mark indexing state fingerprint[%s] as active. Future segments publishes may remediate", fingerprint);
+                                              .distinct()
```
somehow ended up with an extra space here

capistrant commented:
Thank you @clintropolis and @kfaraz for helping get this change reviewed and ready for use. 

@kfaraz if you end up circling back and have thoughts on follow ups, please let me know and I will assess/implement.
